torchao/prototype/sparsity/superblock/blocksparse.py

```diff
@@ -206,6 +241,69 @@ def block_sparse_detach(func, types, args, kwargs):
     )
 
 
+@implements(aten.unsqueeze.default)
+def block_sparse_unsqueeze(func, types, args, kwargs):
+    assert len(args) == 2
+    assert len(kwargs) == 0
+    assert args[-1] == 2
+    bsr = args[0]
+    assert bsr.dim() == 2
+    assert not bsr.requires_grad
+    return BlockSparseTensor(bsr.shape + (1,),
+                             bsr.blocksize, 
+            bsr.crow_indices(),
+            bsr.col_indices(),
+            bsr.values().unsqueeze(-1))
+
+
+@implements(aten.mul.Tensor)
+def block_sparse_mul(func, types, args, kwargs):
+    assert len(args) == 2
+    assert len(kwargs) == 0
+    bsr, t = args
+
+    def my_mul(bsr, t):
+        assert isinstance(bsr, BlockSparseTensor)
+        assert isinstance(t, torch.Tensor)
+        assert bsr.dim() == 3
+        assert t.dim() == 3
+        assert not bsr.requires_grad
+        assert t.size(0) == 1
+        t_blocked = t.view(t.size(0), t.size(1) // 64, 64, 1)
+        masked_t = t_blocked.transpose(0, 1).index_select(0, bsr.col_indices())
+        new_values = bsr.values() * masked_t
+        return BlockSparseTensor(bsr.shape,
+                                 bsr.blocksize,        
+                                 bsr.crow_indices(),
+                                 bsr.col_indices(),
+                                 new_values)
+
+    if isinstance(bsr, torch.Tensor) and isinstance(t, BlockSparseTensor):
+        return my_mul(t, bsr)
+    return my_mul(bsr, t)
+
+
+@implements(aten.sum.dim_IntList)
+def block_sparse_sum(func, types, args, kwargs):
```

Comment:
cc @cpuhrsch This computes the sum properly for the fast path reduction, but doesn't work with compile because of L300 `temp_sum = bsr.values()[start:stop]` which errors out on data dependent flow.

I think we can add a new kernel to the `bsr_dense_addmm` implementation to handle the fast path there instead, and rewrite this using triton

Reply:
Right, so here it's useful to view crow_indices and values as a NestedTensor and then use sum from there :) This is possible because values + crow_indices is like values + offsets.